pkg/yang/find_test.go

```diff
@@ -0,0 +1,290 @@
+				}`,
+			"sys": `
+				submodule sys {
+					belongs-to dev {
+						prefix "d";
+					}
+					include sysdb;
+
+					revision 01-01-01 { description "the start of time"; }
+				}`,
+			"sysdb": `
+				submodule sysdb {
+					belongs-to dev {
+						prefix "d";
+					}
+
+					revision 01-01-01 { description "the start of time"; }
+
+					grouping g { leaf a { type string; } }
+				}`,
+		},
+		inNode: func(ms *Modules) (Node, error) {
+			return FindNode(ms.Modules["dev"], "c")
+		},
+		inName:            "g",
+		wantGroupNodePath: "/sysdb/g",
+	}, {
+		desc: "grouping in indirectly-included submodule with node in submodule",
+		inMods: map[string]string{
+			"dev": `
+				module dev {
+					prefix d;
+					namespace "urn:d";
+					include sys;
+
+					revision 01-01-01 { description "the start of time"; }
+				}`,
+			"sys": `
+				submodule sys {
+					belongs-to dev {
+						prefix "d";
+					}
+					include sysdb;
+
+					revision 01-01-01 { description "the start of time"; }
+
+					container c { leaf b { type string; } }
+				}`,
+			"sysdb": `
+				submodule sysdb {
+					belongs-to dev {
+						prefix "d";
+					}
+
+					revision 01-01-01 { description "the start of time"; }
+
+					grouping g { leaf a { type string; } }
+				}`,
+		},
+		inNode: func(ms *Modules) (Node, error) {
+			return FindNode(ms.SubModules["sys"], "c")
+		},
+		inName:            "g",
+		wantGroupNodePath: "/sysdb/g",
+	}, {
+		desc: "grouping in submodule",
+		inMods: map[string]string{
+			"dev": `
+				module dev {
+					prefix d;
+					namespace "urn:d";
+					import sysdb { prefix "s"; }
+
+					revision 01-01-01 { description "the start of time"; }
+
+					container c { leaf b { type string; } uses s:g; }
+				}`,
+			"sysdb": `
+				module sysdb {
+					prefix sd;
+					namespace "urn:sd";
+
+					revision 01-01-01 { description "the start of time"; }
+
+					grouping g { leaf a { type string; } }
+				}`,
+		},
+		inNode: func(ms *Modules) (Node, error) {
+			return FindNode(ms.Modules["dev"], "c")
+		},
+		inName:            "s:g",
+		wantGroupNodePath: "/sysdb/g",
+	}, {
+		desc: "grouping that uses another grouping both in different modules",
+		inMods: map[string]string{
+			"dev": `
+				module dev {
+					prefix d;
+					namespace "urn:d";
+					import dev2 { prefix "de2"; }
+
+					revision 01-01-01 { description "the start of time"; }
+
+					container c { leaf l { type string; } uses de2:g; }
+				}`,
+			"dev2": `
+				module dev2 {
+					prefix d2;
+					namespace "urn:d2";
+					import dev3 { prefix "de3"; }
+
+					revision 01-01-01 { description "the start of time"; }
+
+					grouping g { leaf a { type string; } uses de3:gg; }
+				}`,
+			"dev3": `
+				module dev3 {
+					prefix d3;
+					namespace "urn:d3";
+
+					revision 01-01-01 { description "the start of time"; }
+
+					grouping gg { leaf b { type string; } }
+				}`,
+		},
+		inNode: func(ms *Modules) (Node, error) {
+			return FindNode(ms.Modules["dev2"], "g")
```

Comment:
If you change this test case to be:

```suggestion
			return FindNode(ms.Modules["dev"], "c")
```

then the test will fail. This is what @aashaikh was running into before and needed the fix referenced in https://github.com/openconfig/goyang/commit/5057f4840240f76faba4f87d84ff51e75597f224

Reply:
Is it correct to search for `gg` from `dev:c` though? It seems to me we shouldn't be demanding `FindGrouping`, which searches from a node, to try to find something from a module it doesn't import from.

I suggested a fix here: https://github.com/openconfig/oc-tools/pull/19#discussion_r778459045

Reply:
Good question!  I think this comes down to the semantics of the function, our current docstring reads:

```
// FindGrouping finds the grouping named name in one of the parent node's
// grouping fields, seen provides a list of the modules previously seen
// by FindGrouping during traversal.  If no parent has the named grouping,
// nil is returned. Imported and included modules are also checked.
```

From this I read that if I have a node `b` that has a parent `a` then I should be able to find a grouping defined in the scope of `a` using `b` as the context node. If `b` is a grouping itself and `a` is a module, then all imports could be argued to be in the context of `a`. I think I agree that if the module that defines `a` doesn't import the module that defines the grouping that is being referenced then that should fail (since one couldn't `use` that grouping in that context).

Let's clarify the docstring to say that the context node that is used for `FindGrouping` defines what the addressable elements are according to YANG namespace rules and then add a test that explicitly ensures that this fails. :-)

Reply:
Clarified docstring and added testcase :)